src/server/pps/server/poller.go

```diff
@@ -176,25 +172,17 @@ func (m *ppsMaster) pollPipelines(ctx context.Context) {
 // to CRASHING
 func (m *ppsMaster) pollPipelinePods(ctx context.Context) {
 	if err := backoff.RetryUntilCancel(ctx, backoff.MustLoop(func() error {
-		kubePipelineWatch, err := m.a.env.KubeClient.CoreV1().Pods(m.a.namespace).Watch(
-			ctx,
-			metav1.ListOptions{
-				LabelSelector: metav1.FormatLabelSelector(metav1.SetAsLabelSelector(
-					map[string]string{
-						"component": "worker",
-					})),
-				Watch: true,
-			})
+		watch, cancel, err := m.kd.WatchPipelinePods(ctx)
 		if err != nil {
 			return errors.Wrap(err, "failed to watch kubernetes pods")
 		}
-		defer kubePipelineWatch.Stop()
+		defer cancel()
 	WatchLoop:
 		for {
 			select {
 			case <-ctx.Done():
 				return nil
-			case event := <-kubePipelineWatch.ResultChan():
+			case event := <-watch:
```

Comment:
oh one thing I was curious about here - I think really we should be doing `event, ok := <- watch` and restarting the watch if `!ok` - I think that explains the "stuck in a loop returning empty events" behavior

Reply:
I'm not sure i follow the empty events bit. But stuck in a loop definitely makes sense